certora/specs/HashConsensus.spec

```diff
@@ -0,0 +1,742 @@
+    setFrameConfig@withrevert(e, epochsPerFrame, fastLaneLengthSlots);
+    bool callReverted = lastReverted;
+
+    assert (!hasRoleR && !isAdmin) => callReverted;
+}
+*/
+
+//  3. setFrameConfig() updates epochsPerFrame in a way that either keeps the current reference slot
+//     the same or increases it by at least the minimum of old and new frame sizes
+// Status: Timeout (only first two asserts)
+// https://prover.certora.com/output/80942/4a45e56dd5e844d09b131889b4b1e304/?anonymousKey=f687e64cc3d822894394894b26e24ef75834d7bb
+// Status: Fails (with last assert)
+// https://prover.certora.com/output/80942/83a93731b535479e96b453701aed235d/?anonymousKey=cb9b95302e423b38dc1403c4f848a464c0585e3b
+rule setFrameConfigCorrectness() {
+    env e; env e2;
+    require e.block.timestamp > saneTimeConfig();   // time moves forward (saneTimeConfig is for correct initializing)
+    require e2.block.timestamp > e.block.timestamp;
+
+    // Get state before
+    uint256 refSlot1; uint256 reportProcessingDeadlineSlot1;
+    refSlot1, reportProcessingDeadlineSlot1 = getCurrentFrame(e);
+    
+    uint256 initialEpoch1; uint256 epochsPerFrame1; uint256 fastLaneLengthSlots1;
+    initialEpoch1, epochsPerFrame1, fastLaneLengthSlots1 = getFrameConfig(e);
+
+    uint256 epochsPerFrame; uint256 fastLaneLengthSlots;
+    setFrameConfig(e2, epochsPerFrame, fastLaneLengthSlots);
+
+    // Get state after
+    uint256 refSlot2; uint256 reportProcessingDeadlineSlot2;
+    refSlot2, reportProcessingDeadlineSlot2 = getCurrentFrame(e2);
+
+    uint256 initialEpoch2; uint256 epochsPerFrame2; uint256 fastLaneLengthSlots2;
+    initialEpoch2, epochsPerFrame2, fastLaneLengthSlots2 = getFrameConfig(e2);
+
+    // verify getter returns updated values
+    assert epochsPerFrame == epochsPerFrame2;               // only those asserts cause timeout
+    assert fastLaneLengthSlots == fastLaneLengthSlots2;     // only those asserts cause timeout
+
+    // verify the comment of the function:
+    // keeps the current reference slot the same or
+    // increases it by at least the minimum of old and new frame sizes
+    uint256 slotsPerEpoch; uint256 secondsPerSlot; uint256 genesisTime;
+    slotsPerEpoch, secondsPerSlot, genesisTime = getChainConfig(e2);
+
+    assert refSlot2 >= refSlot1;  // fails
+
+    // assert (epochsPerFrame2 >= epochsPerFrame1) =>
+    //            ( (refSlot1 == refSlot2) || (refSlot2 >= refSlot1 + epochsPerFrame1 * slotsPerEpoch) );
+    
+    // assert (epochsPerFrame2 <= epochsPerFrame1) =>
+    //            ( (refSlot1 == refSlot2) || (refSlot2 >= refSlot1 + epochsPerFrame2 * slotsPerEpoch) );
+}
+
+
+//  4. setFastLaneLengthSlots() works as expected, setting it to zero disables the fast lane subset
+//     verify with getFrameConfig()
+// Status: Pass
+// https://prover.certora.com/output/80942/133f5763d704400b8bf17e08db8d01d6/?anonymousKey=c467fc94b9572fb48f8790f6c5ba5872f188ea1b
+rule setFastLaneLengthSlotsCorrectness() {
```

Comment:
This seems like a particular case of a more general integrity rule, where instead of zero there could have been an arbitrary value.
Why is this case interesting?